bootstrap/controllers/machine_deployment_controller.go

```diff
@@ -0,0 +1,433 @@
+			upgradedMachines++
+			continue
+		}
+
+		if !m.DeletionTimestamp.IsZero() {
+			log.Info("Machine is being deleted", "machine", m.Name)
+			continue
+		}
+
+		if r.isMachineUpgradeFailed(m) {
+			log.Info("Machine upgrade failed for machine, requeuing...", "machine", m.Name)
+			if err := r.markUpgradeFailed(ctx, scope, m); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to mark upgrade as failed: %w", err)
+			}
+
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+
+		if r.isMachineUpgrading(m) {
+			log.Info("Machine is upgrading, requeuing...", "machine", m.Name)
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+
+		// Machine is not upgraded, mark it for upgrade
+		if err := r.markMachineToUpgrade(ctx, scope, m); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to mark machine to upgrade: %w", err)
+		}
+
+		log.Info("Machine marked for upgrade", "machine", m.Name)
+
+		if err := r.markUpgradeInProgress(ctx, scope); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to mark upgrade as in-progress: %w", err)
+		}
+	}
+
+	if upgradedMachines == len(ownedMachines) {
+		if err := r.markUpgradeDone(ctx, scope); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to mark upgrade as done: %w", err)
+		}
+
+		log.Info("All machines are upgraded")
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+}
+
+// markUpgradeInProgress marks the MachineDeployment as in-place upgrade in-progress.
+func (r *MachineDeploymentReconciler) markUpgradeInProgress(ctx context.Context, scope *MachineDeploymentUpgradeScope) error {
+	annotations := scope.MachineDeployment.Annotations
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// clean up
+	delete(annotations, bootstrapv1.InPlaceUpgradeReleaseAnnotation)
+
+	annotations[bootstrapv1.InPlaceUpgradeToAnnotation] = scope.UpgradeTo
+	annotations[bootstrapv1.InPlaceUpgradeStatusAnnotation] = bootstrapv1.InPlaceUpgradeInProgressStatus
+
+	if err := scope.PatchHelper.Patch(ctx, scope.MachineDeployment); err != nil {
+		return fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.recorder.Eventf(
+		scope.MachineDeployment,
+		corev1.EventTypeNormal,
+		bootstrapv1.InPlaceUpgradeInProgressEvent,
+		"In-place upgrade is in-progress",
+		scope.MachineDeployment.Name,
+	)
+	return nil
+}
+
+// markUpgradeDone marks the MachineDeployment as in-place upgrade done.
+func (r *MachineDeploymentReconciler) markUpgradeDone(ctx context.Context, scope *MachineDeploymentUpgradeScope) error {
+	annotations := scope.MachineDeployment.Annotations
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// clean up
+	delete(annotations, bootstrapv1.InPlaceUpgradeToAnnotation)
+
+	annotations[bootstrapv1.InPlaceUpgradeStatusAnnotation] = bootstrapv1.InPlaceUpgradeDoneStatus
+	annotations[bootstrapv1.InPlaceUpgradeReleaseAnnotation] = scope.UpgradeTo
```

Comment:
We should also add this under `MachineDeployment.spec.template.metadata.annotations` which would propagate to `MachineSet.spec.template.metadata.annotations` and lastly propagate to `Machine.annotations`. This is important for scale up/rolling upgrade compatability. 

Also we should remove this annotation from both sections once an upgrade is requested, and add it after completion.

Reply:
Right, will make the changes. But as a question:
For the scale up we will trigger the reconciliation and add the `upgrade-to` to the machines (manually in this reconciler), this should be enough, right? I also tested scale ups and everything works as expected (new machines get annotated and upgraded).
For rolling upgrade, I'm not sure what we should do here... 
Let's say we previously had an in-place upgrade by annotating the `MachineDeployment`. What should happen when the user triggers a rolling upgrade (by changing `version`) while also keeping the in-place upgrade annotation?

Reply:
Worth mentioning that while the current implementation _does_ cover scaling up, it's probably not standard (in terms of k8s and how annotation propagation works) nor aligned with how the `Ck8sConfigReconciler` works (uses `release` to overwrite snap install instructions)
So I'll change this. Thanks for pointing this out.